internals/secrethub/tpl/v2.go

```diff
@@ -24,95 +37,418 @@ func NewV2Parser() Parser {
+					return nil, err
+				}
+				continue
+			}
+			return nil, ErrIllegalVariableCharacter(p.next, p.lineNo, p.columnNo+1)
 		}
-		keys[k] = v
 	}
-	return t.template.Inject(keys)
 }
 
-// Secrets returns a list of paths to secrets that are used in the template.
-func (t secretTemplateV2) Secrets() []string {
-	set := map[string]struct{}{}
-	for _, path := range t.secrets {
-		set[path] = struct{}{}
+// parseSecret parses the contents of a secret tag up to the closing delimiter.
+// parseSecret should be called after the opening delimiter has been read. The next
+// character from the buffer should be the first character of the contents.
+//
+// when parseSecret returns, the next character in the buffer is the first character
+// after the closing delimiter of the secret tag.
+func (p *v2Parser) parseSecret() (node, error) {
+	path := []node{}
+	err := p.readRune()
+	if err == io.EOF {
+		return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+	}
+	if err != nil {
+		return nil, err
+	}
+	for p.next == ' ' {
+		err = p.readRune()
+		if err == io.EOF {
+			return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for {
+		err = p.readRune()
+		if err == io.EOF {
+			return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		switch p.current {
+		case '$':
+			switch p.next {
+			case '{':
+				err = p.readRune()
+				if err == io.EOF {
+					return nil, ErrVariableTagNotClosed(p.lineNo, p.columnNo+1)
+				}
+				if err != nil {
+					return nil, err
+				}
+				variable, err := p.parseVar()
+				if err != nil {
+					return nil, err
+				}
+				path = append(path, variable)
+
+				err = p.readRune()
+				if err == io.EOF {
+					return nil, ErrSecretTagNotClosed(p.lineNo, p.columnNo+1)
+				}
+				if err != nil {
+					return nil, err
+				}
+			default:
+				return nil, ErrIllegalSecretCharacter(p.current, p.lineNo, p.columnNo)
+			}
+		case ' ':
+			err := p.forwardToClosing([]rune("}}"))
+			if err != nil {
+				return nil, ErrIllegalSecretCharacter(p.current, p.lineNo, p.columnNo)
+			}
+			return secret{
+				path: path,
+			}, nil
+		case '}':
+			switch p.next {
+			case '}':
+				return secret{
+					path: path,
+				}, nil
+			default:
+				return nil, ErrIllegalSecretCharacter(p.current, p.lineNo, p.columnNo)
+			}
+		default:
+			if unicode.IsLetter(p.current) || unicode.IsDigit(p.current) || p.current == '_' || p.current == '-' || p.current == '.' || p.current == '/' || p.current == ':' {
+				path = append(path, character(p.current))
+				continue
+			}
+			return nil, ErrIllegalSecretCharacter(p.current, p.lineNo, p.columnNo)
+		}
 	}
+}
 
-	result := make([]string, len(set))
+// forwardToClosing skips all spaces up to the closing delimiter.
+// It returns an error when characters other than spaces occur before the complete
+// closing delimiter occurs.
+func (p *v2Parser) forwardToClosing(delim []rune) error {
```

Comment:
Call it `delimiters` or `delims` please to signal this can be multiple. Also, consider using the `...rune` instead so this is simple for both single and multiple delimieters:

```golang
parser.forwardToClosing('}')
parser.forwardToClosing('}', '}')
```

Reply:
Also, consider renaming the function to something like `skipSpaceToClosingDelimiter` or just `skipTrailingSpace(delimiters ...rune)`

Reply:
4238e0a189fcd4d7fe0d44ebb2df3e0aada1d559

Reply:
The function also forwards to the end of the delimiter. Do you have a suggestion for the name that also reflects this?

Reply:
Replaced the function in 1f363184b128ed5b8ea8347addef7799d305aa4b